### PR TITLE
Add guest-to-request memory headroom ratio

### PR DESCRIPTION
**Pull request by iholder101**
**What this PR does / why we need it**:
Kubevirt computes [[1](https://github.com/kubevirt/kubevirt/blob/v0.59.0-alpha.2/pkg/virt-controller/services/renderresources.go#L272)] the estimation of the total memory overhead needed for infra components inside virt-launcher which are responsible to run the guest (e.g. libvirtd, qemu, etc).

Not only that this overhead currently suffers from known issues and non-accurate calculation which needs to be fixed - this calculation is in essence an educated guess / estimation, and not an accurate calculation. The reason is that even if a careful profiling will take place (which is a very difficult task to do, since the environments on which we would profile makes the results biased), there are still many components we cannot control, e.g. kernel drivers, kernel configuration, inner QEMU buffer allocations, etc.

To solve this problem, we need to both keep improving the overhead estimations, but also provide a solution for the cluster admin to explicitly add some overhead. This is useful in cases where a special configuration is used, or if the cluster-admin chooses to reduce the risk of workloads OOM killed during node pressures in exchange of workloads consuming more memory (therefore less VMIs could be scheduled on a node).

[1] https://github.com/kubevirt/kubevirt/blob/v0.59.0-alpha.2/pkg/virt-controller/services/renderresources.go#L272

**Which issue(s) this PR fixes**:
Fixes https://bugzilla.redhat.com/show_bug.cgi?id=2165618

**Special notes for your reviewer**:
For a reference, you can look at the temporary solution implemented in HCO: https://github.com/kubevirt/hyperconverged-cluster-operator/pull/2206

**Release note**:

```release-note
Add guest-to-request memory headroom ratio.
This can be enabled by setting `kubevirt.spec.configuration.guestToRequestMemoryHeadroomRatio = "1.234"`
```

### CODE REVIEW

File: api/openapi-spec/swagger.json [hunk 1]
```diff
@@ -16853,6 +16853,10 @@
       "description": "EvictionStrategy defines at the cluster level if the VirtualMachineInstance should be migrated instead of shut-off in case of a node drain. If the VirtualMachineInstance specific field is set it overrides the cluster level one.",
       "type": "string"
      },
+     "guestToRequestMemoryHeadroomRatio": {
+      "description": "GuestToRequestMemoryHeadroomRatio can be used to set a ratio between VM's memory guest and the memory allocated to the compute container. A higher ratio means that the VMs would be less compromised by node pressures, but would mean that fewer VMs could be scheduled to a node. If not set, the default is 1.",
```
Comment (fabiand) [inline api/openapi-spec/swagger.json:16857]: Please add a note if the internal overhead calculations are still taken into account or not (ignored)

Comment (iholder101) [reply api/openapi-spec/swagger.json:16857]: Will do.
FYI: they are.

Comment (iholder101) [reply api/openapi-spec/swagger.json]: done

**Comment by iholder101** [comment]
Hey @vladikr, thanks for your review!

> Thanks for this PR @iholder101
> 
> I think we do need such an additional tunable for the admin to increase the VM calculated overhead in the cluster.
> 
> That said, I would expect the additional calculated "headroom" simply be part of the `GetMemoryOverhead` function. That way the overhead calculation will stay consistent, without the need to separately update the requests and limits. When the requests/limits are calculated, virt-controller adds the overhead to it, it could also add the additional admin-provided overhead at the same step. This would also work for the case when the guest overhead should be overcommitted: using the `OvercommitGuestOverhead`.

Thanks, that's a very good point! Thank you!
I've made the required changed :)

I just want to note that now this field controls the overhead itself only, not the ratio between the guest memory and requests. IOW, the previous implementation did `requests = (<guest-memory> + <overhead>)*<ratio>` while the current implementation does `requests = <guest-memory> + <overhead>*<ratio>`. I actually think it's much better this way.

> I'm also not sure about the `guestToRequestMemoryHeadroomRatio` name. When I think about the main purpose of this tunable, I think that it is to allow the admin to provide an additional overhead - knowing that our overhead calculation is "best effort". Would it make sense to rename it to something like `additinalGuestOverhead` or `increaseFuestOverheadRatio`?
> 
> https://github.com/kubevirt/kubevirt/blob/ed8b899e093e78c79bd4053b3ef4b03f8094df02/pkg/virt-controller/services/renderresources.go#L272

We can discuss the name.
The reason I don't like names like you suggested is that the headroom never actually reaches the guest itself. IMO something like `additinalGuestOverhead` implies that the overhead affects the guest somehow, which is somewhat misleading.

WDYT about something like `additinalVirtInfraOverhead`?

**Comment by iholder101** [comment]
Unit tests pass locally
/test pull-kubevirt-unit-test

File: pkg/virt-controller/services/template.go [hunk 2]
```diff
@@ -1304,7 +1304,7 @@ func checkForKeepLauncherAfterFailure(vmi *v1.VirtualMachineInstance) bool {
 }
 
 func (t *templateService) VMIResourcePredicates(vmi *v1.VirtualMachineInstance, networkToResourceMap map[string]string) VMIResourcePredicates {
-	memoryOverhead := GetMemoryOverhead(vmi, t.clusterConfig.GetClusterCPUArch())
+	memoryOverhead := GetMemoryOverhead(vmi, t.clusterConfig.GetClusterCPUArch(), t.clusterConfig.GetConfig().GuestToRequestMemoryHeadroomRatio)
```
Comment (vladikr) [inline pkg/virt-controller/services/template.go:1307]: I wish there was an easy way to simply get the config from within the GetMemoryOverhead function...

**Comment by vladikr** [comment]
> Hey @vladikr, thanks for your review!
> 
> > Thanks for this PR @iholder101
> > I think we do need such an additional tunable for the admin to increase the VM calculated overhead in the cluster.
> > That said, I would expect the additional calculated "headroom" simply be part of the `GetMemoryOverhead` function. That way the overhead calculation will stay consistent, without the need to separately update the requests and limits. When the requests/limits are calculated, virt-controller adds the overhead to it, it could also add the additional admin-provided overhead at the same step. This would also work for the case when the guest overhead should be overcommitted: using the `OvercommitGuestOverhead`.
> 
> That's a very good point! Thank you! I've made the required changed :)

:+1: 

> 
> I just want to note that now this field controls the overhead itself only, not the ratio between the guest memory and requests. IOW, the previous implementation did `requests = (<guest-memory> + <overhead>)*<ratio>` while the current implementation does `requests = <guest-memory> + <overhead>*<ratio>`. I actually think it's much better this way.
> 
> > I'm also not sure about the `guestToRequestMemoryHeadroomRatio` name. When I think about the main purpose of this tunable, I think that it is to allow the admin to provide an additional overhead - knowing that our overhead calculation is "best effort". Would it make sense to rename it to something like `additinalGuestOverhead` or `increaseFuestOverheadRatio`?
> > https://github.com/kubevirt/kubevirt/blob/ed8b899e093e78c79bd4053b3ef4b03f8094df02/pkg/virt-controller/services/renderresources.go#L272
> 
> We can discuss the name. The reason I don't like names like you suggested is that the headroom never actually reaches the guest itself. IMO something like `additinalGuestOverhead` implies that the overhead affects the guest somehow, which is somewhat misleading.
> 
> WDYT about something like `additinalVirtInfraOverhead`?

I understand your point. I didn't actually read it this way so far. To me "Guest Overhead" on the Virtual Machine Instance reads as an overhead caused by the guest... 
My main concern with names is consistency. We already use the term "Guest Overhead" in many places including other tunables `OvercommitGuestOverhead` [1].
If you prefer not to use this term, then we should find something similar that is not confusing.

Does it make sense?

[1] https://kubevirt.io/user-guide/operations/node_overcommit/#overcommit-the-guest-overhead

File: pkg/virt-controller/services/template.go (hunk 2 above)
Comment (iholder101) [reply pkg/virt-controller/services/template.go:1307]: Yeah, it was a bit painful to write..
I'm not sure how we can improve this in a straight forward way, as this function is being called from different containers. We can always fetch Kubevirt with an API call, but that would be costly in terms of performance.

**Comment by iholder101** [comment]
@vladikr changed the name to `AdditinalGuestOverheadRatio`
PTAL

File: api/openapi-spec/swagger.json [hunk 3]
```diff
@@ -16825,6 +16825,10 @@
     "description": "KubeVirtConfiguration holds all kubevirt configurations",
     "type": "object",
     "properties": {
+     "additinalGuestOverheadRatio": {
```
[the remark this answers is not in the document]
Comment (iholder101) [reply api/openapi-spec/swagger.json:16828]: @vladikr and I discussed the name in comments above.
At least to me, it feels a bit misleading that we mention "guest memory", as the guest memory itself is not affected in any way. @vladikr claimed that this is the naming we chose for many other fields, so I guess it's a topic for further refactoring.

I suggested the name `additinalVirtInfraOverhead` above, as it specifically refers to the virt infra overhead - not to the guest memory.

Please share your thoughts :)

[the remark this answers is not in the document]
Comment (fabiand) [reply api/openapi-spec/swagger.json:16828]: I'd definetly include memory - if it is memory specific.

[the remark this answers is not in the document]
Comment (iholder101) [reply api/openapi-spec/swagger.json]: Changed to `additinalGuestMemoryOverheadRatio`.

**Comment by vladikr** [comment]
Thanks!

/approve

**Comment by iholder101** [comment]
> How to validate the parameter? I modified CR and set “additionalGuestMemoryOverheadRatio: 1.3", but it still encounters OOM. Can you provide an example YAML file?

Modifying `kubevirt.spec.configuration.additionalGuestMemoryOverheadRatio` is exactly what needs to be done.

Perhaps 1.3 is not enough for your case?

Some guiding questions:
* How much memory is allocated for the VM (i.e. from the guest perspective)?
* How much memory is allocated to the `compute` container in `virt-launcher` Pod?
* When the container was OOMed - how much memory did it consume?

You can always try setting a higher value and see if it solves it for your case, but I'd recommend to answer the questions above to ensure your get the right values.

In addition, keep in mind that the multiplier (e.g. `1.3`) does not multiply the memory that's configured for the VM, but multiplies the virt-infra memory that's allocated automatically by Kubevirt. For example, let's say that you define a VM with 2GB of memory, and let's say that Kubevirt automatically adds 200M for virt infra. In this case the total memory of the compute container is 2.2G. If you set `additionalGuestMemoryOverheadRatio` to be 1.3, that applies only to the virt infra, so now the container's memory would be: `2GB+1.3*200M = 2GB+260M=2.26G`.

**Comment by iholder101** [comment]
@tghfly I see you set memory limits == requests. Can you explain why do you set memory limits? Memory limits are dangerous, and I wouldn't define them unless there's a good reason to do so, as if you exceed the limit the Pod is going to be OOM killed.

_BTW, another tip: it's recommended to use `spec.domain.memory.guest`, and remove `spec.domain.resources` entirely. In other words: determine the guest's memory and let Kubevirt determine the exact resources of the pod for you._

After you remove the limits, I'd try to see if the `virt-launcher` pod consumes more memory than it requests. On most cases the answer would be no, but if the answer is yes, try setting `additionalGuestMemoryOverheadRatio` to raise the requested amount.

**Comment by iholder101** [comment]
> @iholder101 Thanks，We use "limits == requests" to ensure Guaranteed QoS. And in the NUMA binding and CPU Pin scenarios, it is also necessary to set “limits == requests”.

Fair enough. You're right, a Guaranteed QoS is necessary sometimes.
But as I've said, having limits is dangerous and you have to stand by the values you define.

You can either set a different `spec.domain.memory.guest` and `spec.domain.resources.memory.request` values to manually define the virt overhead, or play with `additionalGuestMemoryOverheadRatio` until you have enough memory to not be killed.

**Comment by xpivarc** [comment]
> @iholder101 Thanks，We use "limits == requests" to ensure Guaranteed QoS. And in the NUMA binding and CPU Pin scenarios, it is also necessary to set “limits == requests”.

@tghfly I think it would be better if you use `DedicatedCPUPlacement` as this will ensure other containers are in Guaranteed QoS as well and the limits/requests will be set for you

**Pull request merged**